middle_end/flambda2/to_cmm/to_cmm_env.ml

```diff
@@ -208,150 +323,407 @@ let get_exn_extra_args env k =
+     consider: *)
+  (*
+   * let x = box_number 1. in
+   * let y = unbox_number x in
+   * let z = y +. 1. in
+   * ...
+   *)
+  (* because we only ever join effects when translating primitives, and that we
+     only use the effects and coeffects from the flambda code, we generate:
+
+     - a binding "x -> box_number 1." with generative effects (as expected)
+
+     - a binding "y -> 1." (because we inlined "x", and cmm_helper functions
+     eliminated the unbox-of-box), but this binding has generative effects,
+     because it effects are computed as : ece(prim:unbox_number) ∪ ece(x), and
+     since "x" has generative effects, we end up with a binding for "y" that has
+     generative effects.
+
+     - when we translate the addition, we explicitly do not consider inlining
+     effectful expressions, and thus we do not inline "y", since it has effects.
+
+     To counteract that, we instead consider that generative effects arising
+     from Delay/"must inline" primitives do not count, and that we consider the
+     binding to be pure.
+
+     CR gbury: this allows to move allocations marked as `Must_inline_once` past
+     function calls (and other effectful expressions), which can break some
+     allocation-counting tests (the same also applies to allocations marked as
+     `Must_inline_and_duplicate` but that is more expected). *)
+  let effs, classification =
+    let classification =
+      To_cmm_effects.classify_by_effects_and_coeffects effs
     in
-    env, binding
-
-let bind_variable0 ?extra env var ~effects_and_coeffects_of_defining_expr:effs
-    ~may_inline ~defining_expr =
-  let env, binding =
-    create_binding ?extra env ~may_inline effs var defining_expr
+    let change_eff_to effs =
+      effs, To_cmm_effects.classify_by_effects_and_coeffects effs
+    in
+    match[@ocaml.warning "-4"] (inline : a inline), classification with
+    | Must_inline_once, Generative_immutable -> change_eff_to Ece.pure
+    | Must_inline_and_duplicate, Generative_immutable ->
+      change_eff_to Ece.pure_can_be_duplicated
+    | _, _ -> effs, classification
   in
-  if may_inline && is_inlinable_box effs ~extra
-  then
-    (* CR-someday lmaurer: This violates our rule about not moving allocations
-       past function calls. We should either fix it (not clear how) or be rid of
-       that rule. *)
-    { env with pures = Variable.Map.add var binding env.pures }
-  else
-    match To_cmm_effects.classify_by_effects_and_coeffects effs with
-    | Pure -> { env with pures = Variable.Map.add var binding env.pures }
-    | Effect -> { env with stages = Effect (var, binding) :: env.stages }
+  let env, binding = create_binding ?extra env ~inline effs var defining_expr in
+  match (inline : a inline) with
+  | Must_inline_and_duplicate ->
+    (* check that the effects and coeffects allow the expression to be
+       duplicated without changing semantics *)
+    (match classification with
+    | Pure | Generative_immutable -> ()
```

Comment:
We've just removed the possibility for `Generative_immutable` to appear here. Either this case should raise a fatal error too, or we should check the classification before the hack.

Reply:
done